kube-client/src/client/client_ext.rs

```diff
@@ -42,9 +45,90 @@ pub struct Cluster;
 /// You can create this directly, or convert `From` a `String` / `&str`, or `TryFrom` an `k8s_openapi::api::core::v1::Namespace`
 pub struct Namespace(String);
 
+/// Referenced object name resolution
+pub trait ObjectRef<K>: ObjectUrl<K> {
+    fn name(&self) -> Option<&str>;
+}
+
+/// Reference resolver for a specified namespace
+pub trait NamespacedRef<K>: ObjectUrl<K> {
+    /// Resolve reference in the provided namespace
+    fn within(&self, namespace: String) -> impl ObjectRef<K>;
+}
+
+impl<K> ObjectUrl<K> for ObjectReference
+where
+    K: Resource,
+{
+    fn url_path(&self) -> String {
+        url_path(
+            &ApiResource::from_gvk(&self.clone().into()),
+            self.namespace.clone(),
+        )
+    }
+}
```

Comment:
Conversion here makes sense, but wondering if will it work for other "objectreference-like things" with less information. In particular, would we be able to handle:

- [`TypedLocalObjectReference`](https://docs.rs/k8s-openapi/latest/k8s_openapi/api/core/v1/struct.TypedLocalObjectReference.html) - namespace implied same as the object we got it from
- [`LocalObjectReference`](https://docs.rs/k8s-openapi/latest/k8s_openapi/api/core/v1/struct.LocalObjectReference.html) - harder, since this needs to get all type data from parent type

not saying we need to, but it's good to know the limits of the design. what would we need to do do `LocalObjectReference`? Would it even make sense?

Reply:
I believe that both of them are quite ambiguous. I pushed a change which tries to support for both, but for `LocalObjectReference` there is a need of a real field which uses it, which can come only from CR. Also the definitions are open to interpretation. TypedLocalObjectReference has an api_group which is confusing, as it is different sometimes from api_version (which is actually group/version always). Forcing some definition there seems a bit questionable to me, need to think it through. Here are some examples:
- as a [group/version](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/e5d625f96415fd44e6399e9c75e2bd985f5a2288/pkg/ingress/enhanced_backend_builder_test.go#L613)
- as a [group](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/35efc1b73b0f6577469bc5fe236ab9085dd24df4/controllers/openstackfloatingippool_controller.go#L171)

I’d like to try if the exported trait can be implemented externally, and if possibly wrapping a LocalObjectReference and implementing the `NamespaceRef` trait for the wrapped type, providing `From` implementation + some `kopium` changes to recognize selectively LocalObjectReference can be a thing.

Reply:
> `LocalObjectReference` there is a need of a real field which uses it

i found one use-case on [`PodSpec::image_pull_secrets`](https://docs.rs/k8s-openapi/latest/k8s_openapi/api/core/v1/struct.PodSpec.html#structfield.image_pull_secrets) which implicitly references `Secret` object as per [kubernetes docs](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#registry-secret-existing-credentials). this is more awkward than I expected because it's not the same object.

> `TypedLocalObjectReference` has an api_group which is confusing

..yeah. if that's the case then it's probably not super important to worry about it. people can write converters from one object ref to the one they want to disambiguate. that said though, i would say the first example there is wrong. apigroup should be just the group, and the usage of `TypedLocalObjectReference` is an indication that people should do discovery to find it.

Reply:
Yeah, I think this is fair. The `scheme` registration I am not sure how it would work, but some user input feels necessary to solve the ambiguity. Happy to look at PRs or help sketch on issues for it (hard to envision it all from my POV now).